update definitions and change tools related to new NIAC ideas

lukaspie opened the pull request:
- [x] Update definitions to latest `fairmat`
- [x] Implement `nameType`
  - [x] `read_nexus`
  - [x] validation
  - [x] NOMAD parser -> #559
- [x] open enumerations
  - [x] `read_nexus`
  - [x] validation
  - [x] NOMAD parser -> #559
- [x] base class inheritance
  - [x] `read_nexus`
  - [x] validation
  - [x] NOMAD parser

rettigl commented:
Also the chemical formula normalization needs to be updated

lukaspie commented:
> Also the chemical formula normalization needs to be updated

We could do that together with #485

lukaspie commented:
NOMAD dependencies compatibility workflow is failing, but will be fixed seperately by #589

rettigl commented:
I believe we should fix all the plugin tests first before merging, to see that this actually works on real life examples

rettigl reviewed (approved):
Mostly looks good and seems to work, apart from issues mentioned in the main thread. We should maybe still add tests for:
- Inheritance
- the various rename options
- a variatic field that is required by name (i.e. AXISNAME[name] satisfying required name)
- identifier example

rettigl commented:
https://github.com/FAIRmat-NFDI/pynxtools/actions/runs/13903876005/job/38927143402
I believe the config is correct now and the issues come from the problems described above.

lukaspie commented on pyproject.toml:
```diff
@@ -36,6 +36,7 @@ dependencies = [
     "lxml>=4.9.1",
     "toposort>=1.10.0",
     "anytree",
+    "pint==0.17",
```
Yes, they had to fix it at some point because of some upstream dependency. There is a [MR](https://gitlab.mpcdf.mpg.de/nomad-lab/nomad-FAIR/-/merge_requests/2387) that updates it in nomad, we can adjust afterwards accordingly.

lukaspie commented:
> Tests with mpes: -Sample/history gets an additional required child ACTIVITY: ![grafik](https://private-user-images.githubusercontent.com/53396064/423689629-e846047b-afc6-46bf-bcb2-2dbb7e0939c0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyODQ3NTcsIm5iZiI6MTc0MjI4NDQ1NywicGF0aCI6Ii81MzM5NjA2NC80MjM2ODk2MjktZTg0NjA0N2ItYWZjNi00NmJmLWJjYjItMmRiYjdlMDkzOWMwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE4VDA3NTQxN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM4YWEzMjA1MzQwMmM1MzEwMTFhOTMwMDIzZDcyNDE4ZDA4ZGE0MWQ2NWFjZGEwY2Y2ODQ0N2U4Y2RlZmZmNDQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZD9n267l20UZucF2ufgjlrWwguPQ4dtL5oi6ir_7dSA)
> 

This was suggested by NIAC to put into https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/base_classes/NXhistory.nxdl.xml#L31. Their idea was that you alway need at least one activity for each history event. That being said, I don't like it too much and really, there should be no `minOccurs` in base classes. We can either make a proposal to change this or we ignore such cases in our code for now. What do you think?

> * Is this renemaing on purpose: [FAIRmat-NFDI/nexus_definitions@`9e29f6a`/contributed_definitions/NXmanipulator.nxdl.xml#L135](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/9e29f6ac94a89b55489f7f090a895b16e7cf3c7e/contributed_definitions/NXmanipulator.nxdl.xml#L135)

Yes, apparently "amperemeter" is not an English word. Was also suggested by [NIAC review](https://github.com/nexusformat/definitions/pull/1424#discussion_r1995377952).

> * NXresolution has been removed from NXintrument. We need to re-add it in NXmpes then, if the NIAC does not like it.

I'll readd it in https://github.com/FAIRmat-NFDI/nexus_definitions/pull/361.

> * What happened to ELECTRON_DETECTOR[detector]/sensor_pixels?

NIAC didn't like it because there were already some pixel-related fields in `NXdetector`, so they asked for these fields to be removed: https://github.com/nexusformat/definitions/pull/1411/commits/96493f83a96012e71f8608e890bdabdf6db86168. If we still need it, we need to discuss with them again.

> * The code here https://github.com/FAIRmat-NFDI/pynxtools/blob/0d0e7e568fdf460553b060b9560452e3d6802ce5/src/pynxtools/dataconverter/validation.py#L198
>    does not work for variadic fields or attributes, that are required and named in the appdef (i.e., data/AXISNAME[energy] will not satisfy a required data/energy)

Indeed, that's a problem. I'll have a look how to fix it.

> * This [FAIRmat-NFDI/nexus_definitions@`9e29f6a`/base_classes/NXcalibration.nxdl.xml#L112](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/9e29f6ac94a89b55489f7f090a895b16e7cf3c7e/base_classes/NXcalibration.nxdl.xml#L112) needs NX_NUMBER and NX_ANY, maybe also here [FAIRmat-NFDI/nexus_definitions@`9e29f6a`/base_classes/NXparameters.nxdl.xml#L30](https://github.com/FAIRmat-NFDI/nexus_definitions/blob/9e29f6ac94a89b55489f7f090a895b16e7cf3c7e/base_classes/NXparameters.nxdl.xml#L30)

You are right. It was anyway suggested to change `NXparameters` during the MPES standardization: https://github.com/nexusformat/definitions/pull/1424#discussion_r1991536728. I suggest we change it on our side in  https://github.com/FAIRmat-NFDI/nexus_definitions/pull/361 and then make the same changes in the NIAC PR.

rettigl commented:
> This was suggested by NIAC to put into https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/base_classes/NXhistory.nxdl.xml#L31. Their idea was that you alway need at least one activity for each history event. That being said, I don't like it too much and really, there should be no `minOccurs` in base classes. We can either make a proposal to change this or we ignore such cases in our code for now. What do you think?

If this was introduced by Niac, I am fine. The sample_preparation in NXmpes satisfies this required group, I just thought this is an error in the validation.

lukaspie commented on src/pynxtools/dataconverter/nexus_tree.py:
```diff
@@ -111,15 +115,18 @@ class NexusNode(NodeMixin):
             The name of the node.
         type (Literal["group", "field", "attribute", "choice"]):
             The type of the node, e.g., xml tag in the nxdl file.
+        name_type (Optional["specified", "any", "partial"]):
```
Removed the optional, see below

lukaspie commented on src/pynxtools/nomad/schema.py:
```diff
@@ -690,11 +693,20 @@ def __create_field(xml_node: ET.Element, container: Section) -> Quantity:
+                pint.errors.UndefinedUnitError,
+                pint.errors.DefinitionSyntaxError,
+            ) as err:
+                raise NotImplementedError(
+                    f"Unit {nx_dimensionality} is not supported for {name}."
+                ) from err
```
No, this is an error in the schema generation.

Using an intentionally wrong unit "asdkasdnasd", we get 
```
ERROR    pynxtools.nomad.sche 2025-03-18T12:55:03 ERROR: {"event": "Fail to generate metainfo.", "exception": "Traceback (most recent call last):\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/packages/pynxtools/src/pynxtools/nomad/schema.py\", line 702, in __create_field\n    quantity = 1 * ureg(nx_dimensionality)\n                   ^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/.venv/lib/python3.12/site-packages/pint/registry.py\", line 1263, in parse_expression\n    return build_eval_tree(gen).evaluate(\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/.venv/lib/python3.12/site-packages/pint/pint_eval.py\", line 102, in evaluate\n    return define_op(self.left)\n           ^^^^^^^^^^^^^^^^^^^^\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/.venv/lib/python3.12/site-packages/pint/registry.py\", line 1264, in <lambda>\n    lambda x: self._eval_token(x, case_sensitive=case_sensitive, **values)\n              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/.venv/lib/python3.12/site-packages/pint/registry.py\", line 1159, in _eval_token\n    {self.get_name(token_text, case_sensitive=case_sensitive): 1}\n     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/.venv/lib/python3.12/site-packages/pint/registry.py\", line 643, in get_name\n    raise UndefinedUnitError(name_or_alias)\npint.errors.UndefinedUnitError: 'asdkasdnasd' is not defined in the unit registry\n\nThe above exception was the direct cause of the following exception:\n\nTraceback (most recent call last):\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/packages/pynxtools/src/pynxtools/nomad/schema.py\", line 977, in __add_section_from_nxdl\n    return __create_class_section(xml_node)\n           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/packages/pynxtools/src/pynxtools/nomad/schema.py\", line 887, in __create_class_section\n    __create_group(xml_node, class_section)\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/packages/pynxtools/src/pynxtools/nomad/schema.py\", line 822, in __create_group\n    __create_field(field, root_section)\n  File \"/home/lukaspie/fairmat/nomad-distro-dev/packages/pynxtools/src/pynxtools/nomad/schema.py\", line 708, in __create_field\n    raise NotImplementedError(\nNotImplementedError: Unit asdkasdnasd is not supported for energy_dispersion__field.", "target_name": "NXmonochromator", "timestamp": "2025-03-18 13:55.03"}
```
But this does not stop the schema generation, it just doesn't create the Quantity in the metainfo.

lukaspie replied:
This was raising beforehand as well, so I think we can keep it as is for now. This should really be handled by using proper unit examples in NeXus anyway.

rettigl replied:
Sure, just wanted to verify this does not break schema generation

lukaspie commented on src/pynxtools/dataconverter/validation.py:
```diff
@@ -161,7 +161,10 @@ def best_namefit_of(name: str, nodes: Iterable[NexusNode]) -> Optional[NexusNode
+                name_partial = node.name_type == "partial"
+
+                score = get_nx_namefit(instance_name, node.name, name_any, name_partial)
                 if score > -1:
```
I don't think we need to do that anymore, since we require the concept name anyway. We just want to see that the given instance_name and the node are compatible, given the nameType rule. So, the score needs to be above -1, but there will only ever be one node that would potentially fit.

rettigl replied:
Ok

rettigl commented:
> NIAC didn't like it because there were already some pixel-related fields in `NXdetector`, so they asked for these fields to be removed: [nexusformat/definitions@96493f8](https://github.com/nexusformat/definitions/commit/96493f83a96012e71f8608e890bdabdf6db86168). If we still need it, we need to discuss with them again

We probalqy don't need this, I will remove it from the config

rettigl commented:
@mkuehbach The tests for pynxtools-em and pynxtools-apm appear to not yet use the new generalized testing framework, and thus potentially might not pick up any issues arrising with the updated definitions and improved verification. Maybe worth changing that at this point.

lukaspie commented:
> Question: Variadic attributes, should it be @AXISNAME_indices[energy_indices], AXISNAME_indices[@energy_indices], or @AXISNAME_indices[@energy_indices]?

I implemented it as `@AXISNAME_indices[@energy_indices]` now.

lukaspie commented:
> Mostly looks good and seems to work, apart from issues mentioned in the main thread. We should maybe still add tests for:
> 
> * Inheritance

We already test this, but not on the base class extension level. Seems to work though based on the XPS and MPES tests.

> * the various rename options

I added a bunch of tests for this, seems to be working just fine

> * a variatic field that is required by name (i.e. AXISNAME[name] satisfying required name)
> * identifier example

I added tests for `identifierNAME[idenfier]` (that works) and for a named identifier `identifier_1`. The latter also works, but what doesn't work is adding `identifier_1/@type` if that is not defined in the application definition (it is reported as undocumented).

This is an overall bigger problem that the inheritance from siblings does not work properly yet for fields and attributes. That is, `identifier_1` does not actually have `identifierNAME` in its inheritance. Currently, we only add inherited fields/attributes if they have the _exact same_ name (the code is [here](https://github.com/FAIRmat-NFDI/pynxtools/blob/update-defs-and-related-tools/src/pynxtools/dataconverter/nexus_tree.py#L217)). Instead, we should add to the inheritance all the entities that match the name, type, etc. I am already workshipping something, but this breaks in other places.

This is also (partly) the reason why "AXISNAME[energy] satisfying required energy" doesn't work yet, as energy does not actuall ycorrectly inherit all the properties of AXISNAME.

For the latter, we also have to be a bit careful. `NXmpes/NXentry/data/energy` has attributes (like `@type`) which AXISNAME doesn't have. In that case, writing `AXISNAME[energy]` into the template is not correct, as this is not really an AXISNAME anymore. There are certainly ways around this, but I would rather implement this in another PR.

lukaspie commented:
@rettigl since this will lead to additional warnings in the mpes reader, we can make the following workaround for now: the testing framework allows you to have a `*_ref_output.txt` together with your test files. There, you can store the expected output of the reader. We then ignore those messages that are captured by the caplog level, but for which we have no know fix. The logic is implemented  [here](https://github.com/FAIRmat-NFDI/pynxtools/blob/update-defs-and-related-tools/src/pynxtools/testing/nexus_conversion.py#L147).

Essentially, the only things we need to change for now is that we change [this line](https://github.com/FAIRmat-NFDI/pynxtools-mpes/blob/main/scripts/regenerate_examples.sh#L12) to
```
dataconverter xarray_saved_small_calibration.h5 config_file.json eln_data.yaml --reader $READER --nxdl $NXDL --output example_eln.nxs &> eln_ref_output.txt
```
and then we pass that file to the test data in the `test_reader.py` (same for the other example).

For the mpes reader, I imagine this can be an intermediate solution until the sibling inheritance works. For the raman reader, among others, this needs to be implemented.

rettigl commented:
> This is an overall bigger problem that the inheritance from siblings does not work properly yet for fields and attributes. That is, `identifier_1` does not actually have `identifierNAME` in its inheritance. Currently, we only add inherited fields/attributes if they have the _exact same_ name (the code is [here](https://github.com/FAIRmat-NFDI/pynxtools/blob/update-defs-and-related-tools/src/pynxtools/dataconverter/nexus_tree.py#L217)). Instead, we should add to the inheritance all the entities that match the name, type, etc. I am already workshipping something, but this breaks in other places.

Is there actually a way to do this correctly? How would you specify which variadic field a field should inherit from, or if at all? I would say we only do this explicit inheritance by name, and otherwise just ignore a concept name for fields/attributes if the corresponding name is defined as well.
I.e. that e.g. also NXdata/DATA[title] should become NXdata/title. We can maybe emitt an info for that case. 
The same behavior should actually be for groups, I would say. E.g. in NXmpes, if you add ENTRY/SAMPLE[data], it should become ENTRY/data, i.e. NXdata, maybe with an info. Or how do you see this?
The last part is currently not working, i.e. ENTRY/SAMPLE[data] is treated as NXsample, and complains that ENTRY/data is missing.

rettigl commented:
> we can make the following workaround for now

As mentioned above, I would simply omitt the concept name if the instance name is literally defined on the same level, and issue an info.

rettigl commented:
> > we can make the following workaround for now
> 
> As mentioned above, I would simply omitt the concept name if the instance name is literally defined on the same level, and issue an info.

After merging https://github.com/FAIRmat-NFDI/nexus_definitions/pull/361, mpes should pass now. One weird thing is that we still get from read_nexus:
DEBUG: @HDF5_version - IS NOT IN SCHEMA
DEBUG: @delay_indices - IS NOT IN SCHEMA
DEBUG: @kx_indices - IS NOT IN SCHEMA
DEBUG: @ky_indices - IS NOT IN SCHEMA
DEBUG: @type - IS NOT IN SCHEMA

ENTRY/data/@energy_indices also does not have a docstring (we did not provide one in NXmpes, maybe should add).

lukaspie commented:
> > This is an overall bigger problem that the inheritance from siblings does not work properly yet for fields and attributes. That is, `identifier_1` does not actually have `identifierNAME` in its inheritance. Currently, we only add inherited fields/attributes if they have the _exact same_ name (the code is [here](https://github.com/FAIRmat-NFDI/pynxtools/blob/update-defs-and-related-tools/src/pynxtools/dataconverter/nexus_tree.py?rgh-link-date=2025-03-19T20%3A47%3A24Z#L217)). Instead, we should add to the inheritance all the entities that match the name, type, etc. I am already workshipping something, but this breaks in other places.
> 
> Is there actually a way to do this correctly? How would you specify which variadic field a field should inherit from, or if at all? 

Well, we want `identifier_1` to inherit from `NXobject/identifierNAME`, right? So that `identifier_1/@type` is documented. The correct way to do this is to first check all possible siblings (based on the inheritance of a field's/attribute's parent) and then filter those where the `name`/`nameType` as well as the `type` match. For the `type`, it means that a field's type must be a subset of the inherited field's type. This is exactly what [this ToDo](https://github.com/FAIRmat-NFDI/pynxtools/blob/master/src/pynxtools/dataconverter/validation.py#L472) is saying.

> I would say we only do this explicit inheritance by name, and otherwise just ignore a concept name for fields/attributes if the corresponding name is defined as well. I.e. that e.g. also NXdata/DATA[title] should become NXdata/title. We can maybe emitt an info for that case. 

Do you wish to replace that in the template then? How would we do this? We cannot really do this in the MultiFormatReader because we don't handle the nxdl files there and so far, the validation only checks, it doesn't actually modify the template.

> The same behavior should actually be for groups, I would say. E.g. in NXmpes, if you add ENTRY/SAMPLE[data], it should become ENTRY/data, i.e. NXdata, maybe with an info. Or how do you see this? The last part is currently not working, i.e. ENTRY/SAMPLE[data] is treated as NXsample, and complains that ENTRY/data is missing.

I see the issue if you have an instance name that matches that of a different class than what you give in the concept name. In my opinion, we should not blindly replace `ENTRY/SAMPLE[data]` by `ENTRY/data`, because the fields/attributes in that group will not match those of `NXmpes/NXentry/data`. For this case, we should probably actually raise an error since this really messes up the whole structure.

lukaspie commented:
> > The same behavior should actually be for groups, I would say. E.g. in NXmpes, if you add ENTRY/SAMPLE[data], it should become ENTRY/data, i.e. NXdata, maybe with an info. Or how do you see this? The last part is currently not working, i.e. ENTRY/SAMPLE[data] is treated as NXsample, and complains that ENTRY/data is missing.
> 
> I see the issue if you have an instance name that matches that of a different class than what you give in the concept name. In my opinion, we should not blindly replace `ENTRY/SAMPLE[data]` by `ENTRY/data`, because the fields/attributes in that group will not match those of `NXmpes/NXentry/data`. For this case, we should probably actually raise an error since this really messes up the whole structure.

In this seperate PR (#591), I made it so that if you give a concept name and a name that matches with a non-variadic name on the same level, a warning is raised and class is written as undocumented. That is `ENTRY/SAMPLE[data]` would still be written, but with the additional warning "Given group name 'SAMPLE' conflicts with the non-variadic name 'specified_group (req)', which should be of type NXdata."

What do you think of this solution? Note that this currently blocks `AXISNAME[energy]` from being used, because energy is not actually an `AXISNAME`. If we have proper inheritance for fields/attributes, we can adjust this again.

lukaspie commented:
> > > we can make the following workaround for now
> > 
> > 
> > As mentioned above, I would simply omitt the concept name if the instance name is literally defined on the same level, and issue an info.
> 
> After merging [FAIRmat-NFDI/nexus_definitions#361](https://github.com/FAIRmat-NFDI/nexus_definitions/pull/361), mpes should pass now. 
One weird thing is that we still get from read_nexus:
DEBUG: @HDF5_version - IS NOT IN SCHEMA
DEBUG: @delay_indices - IS NOT IN SCHEMA
DEBUG: @kx_indices - IS NOT IN SCHEMA
DEBUG: @ky_indices - IS NOT IN SCHEMA
DEBUG: @type - IS NOT IN SCHEMA
> 
> ENTRY/data/@energy_indices also does not have a docstring (we did not provide one in NXmpes, maybe should add).

`DEBUG: @HDF5_version - IS NOT IN SCHEMA` was coming from a bug in the NXroot attributes writing. It is actually `@HDF5_Version` with an uppercase `V`. I fixed it in the mpes branch already, not it is documented.

The `_indices` are actually a bug. Apparently, the axes are not properly assigned to `AXISNAME`, so `AXISNAME_indices` is also not found. Will fix that later, when/if we have a resolution for https://github.com/nexusformat/definitions/pull/1533.

rettigl commented:
> Do you wish to replace that in the template then? How would we do this? We cannot really do this in the MultiFormatReader because we don't handle the nxdl files there and so far, the validation only checks, it doesn't actually modify the template.

I think we don't need to replace anything, just accept that to fulfill a required field in the verification, no?

> I see the issue if you have an instance name that matches that of a different class than what you give in the concept name. In my opinion, we should not blindly replace `ENTRY/SAMPLE[data]` by `ENTRY/data`, because the fields/attributes in that group will not match those of `NXmpes/NXentry/data`. For this case, we should probably actually raise an error since this really messes up the whole structure.

That is maybe the more proper solution, yes.

rettigl commented:
> In this seperate PR (#591), I made it so that if you give a concept name and a name that matches with a non-variadic name on the same level, a warning is raised and class is written as undocumented. That is `ENTRY/SAMPLE[data]` would still be written, but with the additional warning "Given group name 'SAMPLE' conflicts with the non-variadic name 'specified_group (req)', which should be of type NXdata."
> 
> What do you think of this solution? Note that this currently blocks `AXISNAME[energy]` from being used, because energy is not actually an `AXISNAME`. If we have proper inheritance for fields/attributes, we can adjust this again.

I think this is a good idea. But maybe we only merge this after we have fixed the sibling inheritance issue, so that it does not break our tests?

As we added now all relevant axes as named axes anyways, we can also remove the AXISNAME, and avoid the problem for now altogether. It anyways only shows up because of the flexibility with the "*" notation in the mpes reader to add whatever axes are present.

rettigl commented:
Units inside NXcollection are still reported as undocumented:

`The unit, /ENTRY[0.23_mJcm2]/INSTRUMENT[instrument]/DETECTOR[detector]/COLLECTION[detector_parameters]/amplifier_voltage/@units = V, is being written but has no documentation.`

rettigl commented:
My impression is that the validation has become much slower compared to earlier versions, I will try to verify this and see where the time is being spent.

rettigl commented:
I also encountered a few times now the issue, that NX_FLOAT seems a bit restrictive. If I have e.g. from my machine metadata some field as int, which is supposed to be float, I suggest that we do a conversion in that case, rather than issuing a warning.

lukaspie commented:
After discussion with @sanbrock and @mkuehbach, we merge here to get closer to a new release. Will review sibling inheritance as well as several other issues after the release is made.
